test/cpp/end2end/client_lb_end2end_test.cc

```diff
@@ -367,13 +385,31 @@ TEST_F(ClientLbEnd2endTest, RoundRobin) {
     ports.emplace_back(server->port_);
   }
   SetNextResolution(ports);
-  for (size_t i = 0; i < servers_.size(); ++i) {
-    CheckRpcSendOk();
+  // We cannot derive the connection order from the order reported by the
+  // resolver: different polling engines introduce different delays that affect
+  // which subchannel will transition into READY, even for local addresses.
+  // However, given an initial connection order and in the absence of
+  // disruptions, this order will be periodic.
+  std::vector<int> first_connection_order;
+  while (true) {
+    for (size_t i = 0; i < servers_.size(); ++i) {
+      CheckRpcSendOk();
+      UpdateConnectionOrder(servers_, &first_connection_order);
+    }
+    if (first_connection_order.size() == servers_.size()) break;
+    first_connection_order.clear();
   }
-  // One request should have gone to each server.
+  // "first_connection_order" has unique elements by construction. If its length
+  // equals the number of servers, each server must have gotten a request.
+  EXPECT_EQ(first_connection_order.size(), servers_.size());
+  // Subsequent requests maintain the initial connection ordering.
+  ResetCounters();
```

Comment:
I think there's still a potential problem here.  Let's say that the backends become ready in the order 0, 2, 1.  If backend 1 doesn't become ready until after we've hit backend 2 for the first time, then the first loop above will hit the backends in the order 0, 2, 0, 1.  This means that after we exit the first loop, the next request will go to backend 2, not backend 0.  However, the second loop below assumes that the first request it sends will go to backend 0.

I think that right after this line, we need to do something like this to reset round_robin to the start of the backend list:

```
WaitForServer(servers_.size() - 1);
```

Reply:
Done